File: lab5.2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <algorithm>
#include <ctime>
#include <cstdlib>

using namespace std;

struct Visitor{
    string ticket; //номер талона, время
    int duration;
};

string generate_tick()
{
    int num = rand()%1000; // берёт остаток от деления на 1000 взятый по модулю
    return "T" + to_string(num); //добавляем T к преобразованной в строку переменной num
}

int main()
{
    srand(time(0));
    int num_win;
    cout << "Введите количество окон: ";
    cin >> num_win;
    if(num_win >=0){
    vector<Visitor> visitors; // очередь всех посетителей
    vector<vector<Visitor>> windows(num_win); //посетители по каждому окну
    vector<int> win_times(num_win, 0); //время обслуживания в каждом окне

    string cmd;
    cout << "Введите команду: ";
    while(cin >> cmd){

        if(cmd == "ENQUEUE"){
            int dur;
            cin >> dur; //ввод продолжительности
            if(dur >0){
            visitors.push_back({generate_tick(),dur}); //добавление посетителя в очередь и генерация его билета

            cout << "-> " << visitors.back().ticket << endl; //вывод номера
            }
            else {
                cout << "Некорректное значение" << endl;
            }
        }
    
    else if(cmd == "DISTRIBUTE"){
        for (const auto& v : visitors) {
            int min_index = min_element(win_times.begin(), win_times.end()) - win_times.begin();
        //поиск окна с минимальной суммой времени

            windows[min_index].push_back(v); //в это окно назначаем посетителя
            win_times[min_index] += v.duration; // нагрузку увеличиваем
        }
        for (int i = 0; i < num_win; ++i) { //проход по всем онкнам от 0 до количества окон
            cout << "Окно " << (i + 1) << " " << win_times[i] << " минут: "; //вывод информацию о всех окон
        
            if (windows[i].empty()) { //если посетителя в окне нету
                cout << "(пусто)";

            } 
            else { //если окно не пустое, проходимся по посетителям это окне
                for (auto j = windows[i].begin(); j != windows[i].end(); j++) {
                    //итератор на начало списка и проход до конца
                    cout << (*j).ticket; //разыменываем, получаем доступ и выводим поле ticket
                    if (next(j) != windows[i].end()) 
                        cout << ", "; //если следующей элемент не равен последнему, запятая
                }
            }
            cout << "\n";
        }
        break;
    }
    else if(cmd == "INFO")
    {
        cout << "Доступные команды: ENQUEUE DISTRIBUTE INFO" << endl;
    }
    else{
        cout << "Неизвестная команда" << endl;
        continue;
        }
    }
}else{
    cout << "Некорректное значение окон" << endl;
}

    return 0;
}

File: lab5.2.go
package main

import (
    "fmt"
    "math/rand"
    "strings"
    "time"
)

type Visitor struct {
    ticket   string
    duration int
}

func generateTicket() string {
    return fmt.Sprintf("T%d", rand.Intn(1000))
}

func main() {
    rand.Seed(time.Now().UnixNano())

    var numWin int
    fmt.Print("Введите количество окон: ")
    _, err := fmt.Scan(&numWin)
    if err != nil || numWin <= 0 {
        fmt.Println("Некорректное значение окон")
        return
    }

    var visitors []Visitor
    windows := make([][]Visitor, numWin)
    winTimes := make([]int, numWin)

    var cmd string
    fmt.Print("Введите команду: ")
    for {
        _, err := fmt.Scan(&cmd)
        if err != nil {
            break
        }
        cmd = strings.ToUpper(cmd)

        switch cmd {
        case "ENQUEUE":
            var dur int
            fmt.Scan(&dur)
            if dur > 0 {
                ticket := generateTicket()
                visitors = append(visitors, Visitor{ticket, dur})
                fmt.Printf("-> %s\n", ticket)
            } else {
                fmt.Println("Некорректное значение")
            }

        case "DISTRIBUTE":
            for _, v := range visitors {
                minIndex := 0
                for i := 1; i < numWin; i++ {
                    if winTimes[i] < winTimes[minIndex] {
                        minIndex = i
                    }
                }
                windows[minIndex] = append(windows[minIndex], v)
                winTimes[minIndex] += v.duration
            }

            for i := 0; i < numWin; i++ {
                fmt.Printf("Окно %d %d минут: ", i+1, winTimes[i])
                if len(windows[i]) == 0 {
                    fmt.Println("(пусто)")
                } else {
                    var names []string
                    for _, v := range windows[i] {
                        names = append(names, v.ticket)
                    }
                    fmt.Println(strings.Join(names, ", "))
                }
            }
            return

        case "INFO":
            fmt.Println("Доступные команды: ENQUEUE DISTRIBUTE INFO")

        default:
            fmt.Println("Неизвестная команда")
        }
        fmt.Print("Введите команду: ")
    }
}